api/apps/api/src/modules/scenario-specification/scenario-specification.ts

```diff
@@ -0,0 +1,264 @@
+type SpecPublishFailure = typeof DraftNotPublished | typeof SpecMissingFeatures;
+
+class Specification {
+  private constructor(
+    public readonly id: string,
+    private configuration: FeatureConfig[] = [],
+    private draft: boolean,
+    private activated: boolean,
+  ) {
+    //
+  }
+
+  static from(snapshot: SpecificationSnapshotInput): Specification {
+    return new Specification(
+      snapshot.id,
+      snapshot.config,
+      snapshot.draft,
+      snapshot.activated,
+    );
+  }
+
+  static new(input: FeatureConfigInput[], draft: boolean): Specification {
+    return new Specification(
+      v4(),
+      input.map((input) => ({
+        ...input,
+        featuresDetermined: false,
+        resultFeatures: [],
+      })),
+      draft,
+      false,
+    );
+  }
+
+  toSnapshot(): SpecificationSnapshot {
+    return {
+      id: this.id,
+      draft: this.draft,
+      readyToActivate:
+        this.allFeaturesDetermined() && this.allFeaturesCalculated(),
+      config: this.configuration,
+      activated: this.activated,
+      featuresDetermined: this.allFeaturesDetermined(),
+    };
+  }
+
+  publish(): Either<SpecPublishFailure, SpecPublishSuccess> {
+    if (this.draft) {
+      return left(DraftNotPublished);
+    }
+
+    if (!this.allFeaturesDetermined()) {
+      return left(SpecMissingFeatures);
+    }
+
+    if (this.allFeaturesCalculated()) {
+      return right(SpecCanBeActivated);
+    }
+
+    return right(SpecCanBePublished);
+  }
+
+  activate(): boolean {
+    if (this.allFeaturesCalculated() && this.allFeaturesDetermined()) {
+      this.activated = true;
+    }
+    return this.activated;
+  }
+
+  determineFeatures(features: FeaturesFound['spec']) {
+    features.forEach((feature) => {
+      const configPiece = this.configuration.find(
+        (featureCfg) =>
+          featureCfg.baseFeatureId === feature.baseFeatureId &&
+          featureCfg.againstFeatureId === feature.againstFeatureId,
+      );
+      if (!configPiece) {
+        return;
+      }
+      configPiece.resultFeatures = feature.features;
+      configPiece.featuresDetermined = true;
+    });
+  }
+
+  markAsCalculated(
+    featureIds: string[],
+  ): typeof SpecCanBeActivated | undefined {
+    this.configuration.forEach((featureConfig) =>
+      featureConfig.resultFeatures.forEach((feature) => {
+        if (featureIds.includes(feature.id)) {
+          feature.calculated = true;
+        }
+      }),
+    );
+
+    return this.allFeaturesCalculated() ? SpecCanBeActivated : undefined;
+  }
+
+  allFeaturesDetermined(): boolean {
+    return this.configuration.every(
+      (featureConfig) => featureConfig.featuresDetermined,
+    );
+  }
+
+  allFeaturesCalculated(): boolean {
+    return this.configuration.every((featureConfig) =>
+      featureConfig.resultFeatures.reduce<boolean>(
+        (calculated, feature) => calculated && feature.calculated,
+        true,
+      ),
+    );
```

Comment:
```ts
return this.configuration.flatMap(configuration => configuration.resultFeatures).every(feature => feature.calculated)
```

Reply:
fixed in https://github.com/Vizzuality/marxan-cloud/pull/371/commits/4dbad00ade491ccca947032a295834751eab9b2d